java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java

```diff
@@ -325,24 +325,36 @@ public ProcessProofRules invoke(GlobalContext rewritingContext, long initCounter
                 termContext.setKOREtoBackendKILConverter(converter);
                 javaRules = rules.stream()
                         .map(r -> converter.convert(Optional.<Module>empty(), r))
-                        .map(r -> new org.kframework.backend.java.kil.Rule(
-                                r.label(),
-                                evaluate(r.leftHandSide(), r.getRequires(), termContext),
-                                evaluate(r.rightHandSide(), r.getRequires(), termContext),
-                                r.requires(),
-                                r.ensures(),
-                                r.freshConstants(),
-                                r.freshVariables(),
-                                r.lookups(),
-                                r.att(),
-                                termContext.global()))
+                        .map(this::evaluateRule)
                         .collect(Collectors.toList());
                 return this;
             }
 
+            public org.kframework.backend.java.kil.Rule evaluateRule(org.kframework.backend.java.kil.Rule rule) {
+                //We need this ConsTerm only to evaluate the constraint. That's why we use an empty first argument.
+                ConstrainedTerm constraintHolder = new ConstrainedTerm(
+                        ConjunctiveFormula.of(termContext.global()),
+                        rule.getRequires(),
+                        termContext).expandPatterns(true);
+
+                ConjunctiveFormula constraint = constraintHolder.constraint();
```

Comment:
could you explain what's the difference between `constraint` and `rule.getRequires()`?

Reply:
The constraint is "fully evaluated and simplified" while `rule.getRequires()` is not. 2 ddifferences: 1. expandPatterns() can add new constraints, but this is not used in specifications that I know. 2. It is simplified. This is most important part, because it evaluates substitutions.

Reply:
Oh, now I see that `expandPatterns()` is applied to `rule.getRequires()`. Sounds good!